pkg/build/controller/image_change_controller.go

```diff
@@ -81,18 +81,42 @@ func (c *ImageChangeController) HandleImageRepo(repo *imageapi.ImageRepository)
 
 		if shouldBuild {
 			glog.V(4).Infof("Running build for buildConfig %s in namespace %s", config.Name, config.Namespace)
-			b := buildutil.GenerateBuildFromConfig(config, nil, subs)
-			if err := c.BuildCreator.Create(config.Namespace, b); err != nil {
-				return fmt.Errorf("error starting build for buildConfig %s: %v", config.Name, err)
+			// trigger the generation of a new build
+			request := &buildapi.BuildRequest{ObjectMeta: kapi.ObjectMeta{Name: config.Name}}
+			if _, err := c.BuildConfigInstantiator.Instantiate(config.Namespace, request); err != nil {
+				return fmt.Errorf("Error instantiating build from config %s: %v", config.Name, err)
 			}
-			if err := c.BuildConfigUpdater.Update(config); err != nil {
-				// This is not a retryable error because the build has been created.  The worst case
-				// outcome of not updating the buildconfig is that we might rerun a build for the
-				// same "new" imageid change in the future, which is better than running the build
-				// 2+ times by retrying it here.
-				return ImageChangeControllerFatalError{Reason: fmt.Sprintf("Error updating buildConfig %s with new LastTriggeredImageID", config.Name), Err: err}
+			// and update the config
+			if err := c.updateConfig(config); err != nil {
+				// This is not a retryable error. The worst case outcome of not updating the buildconfig
+				// is that we might rerun a build for the same "new" imageid change in the future,
+				// which is better than guaranteeing we run the build 2+ times by retrying it here.
+				return ImageChangeControllerFatalError{
+					Reason: fmt.Sprintf("Error updating buildConfig %s with new LastTriggeredImageID", config.Name),
+					Err:    err,
+				}
 			}
 		}
 	}
 	return nil
 }
+
+// updateConfig is responsible for updating current BuildConfig object which was changed
+// during instantiate call, it basically copies LastTriggeredImageID to fresh copy
+// of the BuildConfig object
+func (c *ImageChangeController) updateConfig(config *buildapi.BuildConfig) error {
+	item, _, err := c.BuildConfigStore.Get(config)
```

Comment:
It's not reducing risk. At this point we ALWAYS hit the collision, because BuildConfig is being updated inside the generator (LastVersion is incremented), thus we need to get fresh copy of it and copy the triggers.